pkg/controller/nodekeeper/nodekeeper_controller.go

```diff
@@ -0,0 +1,225 @@
+package nodekeeper
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/go-logr/logr"
+	configv1 "github.com/openshift/api/config/v1"
+	machineconfigapi "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	upgradev1alpha1 "github.com/openshift/managed-upgrade-operator/pkg/apis/upgrade/v1alpha1"
+	"github.com/openshift/managed-upgrade-operator/pkg/metrics"
+	"github.com/openshift/managed-upgrade-operator/pkg/osd_cluster_upgrader"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("controller_nodekeeper")
+
+var (
+	pdbForceDrainTimeout  int32
+	upgradeConfigNotfound = errors.New("Currently no UpgradeConfig")
+)
+
+// Add creates a new NodeKeeper Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileNodeKeeper{
+		client:               mgr.GetClient(),
+		scheme:               mgr.GetScheme(),
+		metricsClientBuilder: metrics.NewBuilder(),
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("nodekeeper-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource UpgradeConfig, status change will not trigger a reconcile
+	err = c.Watch(&source.Kind{Type: &corev1.Node{}}, &handler.EnqueueRequestForObject{})
```

Comment:
You can add a predicate in here to only watch non master nodes. There is an example predicate in upgradeconfig controller:
https://github.com/openshift/managed-upgrade-operator/blob/ac2c24dcb55cd90ce8dab2e7395b0b27e81b0d3e/pkg/controller/upgradeconfig/upgradeconfig_controller.go#L61

A master has a label `"node-role.kubernetes.io/master": ""` used here:
https://github.com/openshift/managed-upgrade-operator/blob/ac2c24dcb55cd90ce8dab2e7395b0b27e81b0d3e/pkg/scaler/machineSetScaler.go#L183

Reply:
resolved: 18a3825c33a3c99da7b8c7b5dd26992128533a46